include/seqan3/alignment/cigar_conversion.hpp

```diff
@@ -0,0 +1,392 @@
+                               + std::to_string(std::ranges::size(query))};
+
+    // Assign the sequence to the alignment (a tuple of 2 gap decorators)
+    // -------------------------------------------------------------------------
+    using gapped_reference_type = gap_decorator<decltype(reference | views::slice(0, 0))>;
+    using gapped_sequence_type = gap_decorator<decltype(query | views::slice(0, 0))>;
+    using alignment_type = std::tuple<gapped_reference_type, gapped_sequence_type>;
+
+    alignment_type alignment{};
+
+    assign_unaligned(get<0>(alignment),
+                     reference
+                         | views::slice(zero_based_reference_start_position,
+                                        zero_based_reference_start_position + reference_length));
+    assign_unaligned(get<1>(alignment), query | views::slice(soft_clipping_start, soft_clipping_start + query_length));
+
+    // Insert gaps into the alignment based on the cigar vector
+    // -------------------------------------------------------------------------
+    using std::get;
+    auto current_ref_pos = std::ranges::begin(get<0>(alignment));
+    auto current_read_pos = std::ranges::begin(get<1>(alignment));
+
+    for (auto [cigar_count, cigar_operation] : cigar_vector)
+    {
+        assert(('S'_cigar_operation == cigar_operation) || ('H'_cigar_operation == cigar_operation)
+               || ('M'_cigar_operation == cigar_operation) || ('='_cigar_operation == cigar_operation)
+               || ('X'_cigar_operation == cigar_operation) || ('D'_cigar_operation == cigar_operation)
+               || ('N'_cigar_operation == cigar_operation) || ('I'_cigar_operation == cigar_operation)
+               || ('P'_cigar_operation == cigar_operation));
+
+        if (('M'_cigar_operation == cigar_operation) || ('='_cigar_operation == cigar_operation)
+            || ('X'_cigar_operation == cigar_operation))
+        {
+            std::ranges::advance(current_ref_pos, cigar_count);
+            std::ranges::advance(current_read_pos, cigar_count);
+        }
+        else if (('D'_cigar_operation == cigar_operation) || ('N'_cigar_operation == cigar_operation))
+        {
+            // insert gaps into read
+
+            assert(std::distance(current_read_pos, std::ranges::end(get<1>(alignment))) >= 0);
+            current_read_pos = get<1>(alignment).insert_gap(current_read_pos, cigar_count);
+            ++current_read_pos;
+            std::ranges::advance(current_ref_pos, cigar_count);
+        }
+        else if (('I'_cigar_operation == cigar_operation)) // Insert gaps into ref
+        {
+            assert(std::ranges::distance(current_ref_pos, std::ranges::end(get<0>(alignment))) >= 0);
+            current_ref_pos = get<0>(alignment).insert_gap(current_ref_pos, cigar_count);
+            ++current_ref_pos;
+            std::ranges::advance(current_read_pos, cigar_count);
+        }
+        else if (('P'_cigar_operation == cigar_operation)) // skip padding
```

Comment:
Why no asserts here?

Reply:
I assert at the start of the for loop so i don't need to assert here? not sure if I understand what you mean.

Reply:
I meant these asserts for the other cases where you insert gaps:
```
assert(std::ranges::distance(current_ref_pos, std::ranges::end(get<0>(alignment))) >= 0);
```